File: 10-10-2023_Minimum_Number_of_operations_to_make_array_continuous.cpp
class Solution {
public:
    int minOperations(vector<int>& nums){
        int n = nums.size(), ans = n;
        set<int> s;
        for(int i : nums){
            s.insert(i);
        }
        vector<int> unique_elements(s.begin(), s.end());
        int size = unique_elements.size();
        for(int start = 0; start < size; start++){
            int end_index = upper_bound(unique_elements.begin() + start, unique_elements.end(), unique_elements[start] + n - 1) - unique_elements.begin();
            ans = min(ans, n - (end_index - start));
        }
        return ans;
    }
};

File: 10-10-2023_Minimum_Number_of_operations_to_make_array_continuous.java
class Solution{
    public int minOperations(int[] nums) {
        int n = nums.length, ans = n;
        Set<Integer> s = new HashSet<>();
        for (int num : nums) {
            s.add(num);
        }
        List<Integer> unique_elements = new ArrayList<>(s);
        Collections.sort(unique_elements);
        int size = unique_elements.size();
        for (int start = 0; start < size; start++) {
            int end_index = upperBound(unique_elements, start, size, unique_elements.get(start) + n - 1);
            ans = Math.min(ans, n - (end_index - start));
        }
        return ans;
    }
    private static int upperBound(List<Integer> list, int low, int high, int element) {
        while (low < high) {
            int middle = low + (high - low) / 2;
            if (element >= list.get(middle)) {
                low = middle + 1;
            } else {
                high = middle;
            }
        }
        return low;
    }
}
